test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DollarApplyGroupByTests.cs

```diff
@@ -0,0 +1,1717 @@
+        {
+            var queryable = this.dsContext.CreateQuery<Sale>(salesEntitySetName);
+
+            var aggregateQuery = queryable.GroupBy(
+                d1 => d1.Product.Category.Name,
+                (d1, d2) => new SalesGroupedResult05(d2.Average(d3 => d3.Amount))
+                {
+                    CategoryName = d1
+                });
+
+            Assert.Equal(
+                string.Format("{0}/Sales?$apply=groupby((Product/Category/Name),aggregate(" +
+                "Amount with average as averageAmount))", serviceUri),
+                aggregateQuery.ToString());
+
+            MockGroupBy_ConstructorInitialization();
+
+            var aggregateResult = Assert.Single(aggregateQuery.ToArray());
+
+            Assert.Equal("Food", aggregateResult.CategoryName);
+            Assert.Equal(4M, aggregateResult.AverageAmount);
+        }
+
+        [Fact]
+        public void GroupByResultSelector_OnFilteredInputSet_ExpressionTranslatedToExpectedUri()
+        {
+            // Arrange
+            var queryable = this.dsContext.CreateQuery<Sale>(salesEntitySetName);
+
+            var query = queryable.Where(d => d.CurrencyCode.Equals("USD"))
+                .GroupBy(d1 => new { d1.Product.Color }, (d1, d2) => new
+                {
+                    ProductAvgTaxRate = d2.Average(d3 => d3.Product.TaxRate)
+                });
+
+            // Act & Assert
+            var expectedAggregateUri = $"{serviceUri}/{salesEntitySetName}?$apply=filter(CurrencyCode eq 'USD')" +
+                $"/groupby((Product/Color),aggregate(Product/TaxRate with average as ProductAvgTaxRate))";
+            Assert.Equal(expectedAggregateUri, query.ToString());
+        }
+
+        #region Mock Aggregation Responses
+
+        private void MockGroupBy_Sum_ByConstant()
+        {
+            string mockResponse = string.Format("{{" +
+                "\"@odata.context\":\"{0}/$metadata#Numbers(" +
+                "SumIntProp,SumNullableIntProp," +
+                "SumDoubleProp,SumNullableDoubleProp," +
+                "SumDecimalProp,SumNullableDecimalProp," +
+                "SumLongProp,SumNullableLongProp," +
+                "SumSingleProp,SumNullableSingleProp)\"," +
+                "\"value\":[" +
+                "{{\"@odata.id\":null," +
+                "\"SumIntProp\":506,\"SumNullableIntProp\":530," +
+                "\"SumDoubleProp\":464.72,\"SumNullableDoubleProp\":534.02," +
+                "\"SumDecimalProp\":559.4,\"SumNullableDecimalProp\":393.7," +
+                "\"SumLongProp\":1298,\"SumNullableLongProp\":993," +
+                "\"SumSingleProp\":333.79,\"SumNullableSingleProp\":528.44}}" +
+                "]}}", serviceUri);
+
+            InterceptRequestAndMockResponse(mockResponse);
+        }
+
+        private void MockGroupBy_Sum_BySingleProperty()
+        {
+            string mockResponse = string.Format("{{" +

```

Comment:
Out of curiosity, how does the client determine the right project plan compiler to use for this response, since you created a new `GroupByProjectionPlanCompiler`?

Reply:
@habbes This happens in the `ODataEntityMaterializer`. The logic looks like this:
```csharp
// The KeySelectorMap property is initialized and populated with a least one item if we're dealing with a GroupBy expression.
if (queryComponents.GroupByKeySelectorMap?.Count > 0)
{
    result = GroupByProjectionPlanCompiler.CompilePlan(projection, queryComponents.NormalizerRewrites, queryComponents.GroupByKeySelectorMap);
}
else
{
    result = ProjectionPlanCompiler.CompilePlan(projection, queryComponents.NormalizerRewrites);
}
```